modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java

```diff
@@ -110,23 +109,27 @@ private void requestStart(ChannelHandlerContext ctx) {
             // this looks like a malformed request and will forward without validation
             ctx.channel().eventLoop().submit(() -> forwardFullRequest(ctx));
         } else {
-            validator.apply(httpRequest, ctx.channel(), new ActionListener<>() {
-                @Override
-                public void onResponse(Void unused) {
-                    // Always use "Submit" to prevent reentrancy concerns if we are still on event loop
-                    ctx.channel().eventLoop().submit(() -> forwardFullRequest(ctx));
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    // Always use "Submit" to prevent reentrancy concerns if we are still on event loop
-                    ctx.channel().eventLoop().submit(() -> forwardRequestWithDecoderExceptionAndNoContent(ctx, e));
-                }
-            });
+            assert threadContext.isDefaultContext();
+            // this prevents thread-context changes to propagate to the validation listener
+            // atm the validation listener submits to the event loop executor, which doesn't know about the ES thread-context,
+            // so this is just a defensive play, in case the code inside the listener changes to not use the event loop executor
+            ContextPreservingActionListener<Void> contextPreservingListener = new ContextPreservingActionListener<>(
+                threadContext.wrapRestorable(threadContext.newStoredContext()),
+                ActionListener.wrap(aVoid ->
+                // Always use "Submit" to prevent reentrancy concerns if we are still on event loop
+                ctx.channel().eventLoop().submit(() -> forwardFullRequest(ctx)), e ->
+                // Always use "Submit" to prevent reentrancy concerns if we are still on event loop
+                ctx.channel().eventLoop().submit(() -> forwardRequestWithDecoderExceptionAndNoContent(ctx, e)))
+            );
+            // this prevents thread-context changes to propagate beyond the validation, as netty worker threads are reused
+            try (ThreadContext.StoredContext ignore = threadContext.newStoredContext()) {
+                validator.apply(httpRequest, ctx.channel(), contextPreservingListener);
+            }
```

Comment:
I've incorporated the above changes in a PR of their own, see https://github.com/elastic/elasticsearch/pull/95498 . I propose we merge that one first.